### DISCUSSION TITLE
Water Polygons display

### DISCUSSION THREAD

**Discussion by BenOnTrack** [systemed/tilemaker#525] [Q&A]
Hi Richard and co,

For the first time I'd now like to integrate the water polygons. The README is clear, and according to tilemaker logs everything has been correctly generated. I've kept the config 

`		"ocean":{ "minzoom": 0,  "maxzoom": 14, "source": "coastline/water_polygons.shp", "simplify_below": 13, "simplify_level": 0.0001, "simplify_ratio": 2, "write_to": "water", "index": true }`

and I do have a "water" layer filled.

I'd like to understand which tags the shapefile is holding and how it got filled into the layer `ocean` - is there some code that should stand in the process.lua ? Is this processed via the `attribute_function` ? I don't see any `way:Layer("ocean")` in any example, such as process-example, nor process-openmaptiles. I'm just trying to find out:
- which data/table of the shapefile is processed
- where
- how it ends up in a layer 

I'm using maplibre-gl, and I have the following style

```
{
      "id": "water",
      "type": "fill",
      "metadata": {"mapbox:group": "1444849382550.77"},
      "source": "openmaptiles",
      "source-layer": "water",
      "filter": [
        "all",
        ["!=", "intermittent", "yes"],
        ["!=", "brunnel", "tunnel"]
      ],
      "layout": {"visibility": "visible"},
      "paint": {"fill-color": "hsl(210, 67%, 85%)"}
    },
    {
      "id": "water-intermittent",
      "type": "fill",
      "metadata": {"mapbox:group": "1444849382550.77"},
      "source": "openmaptiles",
      "source-layer": "water",
      "filter": [ "all", ["==", "intermittent", "yes"] ],
      "layout": {"visibility": "visible"},
      "paint": {"fill-color": "hsl(210, 67%, 85%)", "fill-opacity": 0.7}
    }

```

but my map doesn't display anything beyond my pbf source (New Zealand extract) while I would have expected the rest of the map to be blue, and the coastlines for each countries.

Curious what I'm doing wrong...
Thanks in advance for your support!
Ben

**Comment by systemed** [comment]
Shapefile data is processed according to the layers you specify in the JSON file, and (if you supply one) `attribute_function`.

tilemaker takes the bounding box from the .osm.pbf file, and gives you an .mbtiles which covers that area. If you want a bigger area - e.g. you want to have coastline data for the rest of the world - then you can change the bounding box with the `--bbox` argument to tilemaker. Bear in mind though that this will generate tiles up to z14 for the whole world!

If you want to generate just (say) tiles up to z6 for the whole world, but tiles up to z14 for New Zealand, then you could do that by running tilemaker once to generate the ocean tiles with appropriate json/lua config, then using `--merge` to add the New Zealand data into that.

**Comment by BenOnTrack** [comment]
Thanks

The  attribute_function(attr,layer) takes 2 arguments:
- `attr`
- `layer`

How does it then work?

Is it called each time a shapefile is referenced in the config file, and the` layer` is the one declared in the config file ? 

OR

Is it supposed to be called in the process.lua (within ` way_function`) to be run ?  

But then, in any case, what about the `attr` ?  

I might be missing some documentation, or not looking at the right place ?

My initial goal is to generate NZ up to zoom 14 and the entire world at a lower zoom level (something similar to what's provided by openmaptiles on their website)

**Reply by systemed** [reply]
`attribute_function` is documented in https://github.com/systemed/tilemaker/blob/master/docs/CONFIGURATION.md#shapefiles. It's called for each shapefile geometry that tilemaker reads, just as `way_function` is called for each OSM way that tilemaker reads.

`layer` is the layer name as declared in your JSON file. `attr` is a Lua table (hash) of the attributes read from the shapefile. You return a Lua table (hash) of the vector tile attributes you want to set for that object.

**Reply by BenOnTrack** [reply]
A bbox is a rectangle, right ? So if I still don't see any of my background turning into my "ocean" color, it means that I still have an issue, but most likely with my style (on the rendered side) and not with the tiles.

At the moment, I'm using the following style, to render the tiles processed with the `process-openmaptiles.lua` 

```
{
      "id": "ocean",
      "type": "fill",
      "source": "openmaptiles",
      "source-layer": "ocean",
      "layout": {"visibility": "visible"},
      "paint": {"fill-color": "hsl(210, 67%, 85%)"}
    }
```

I'm using the current `attribute_function`, just to add a class to my layer

```
function attribute_function(attr,layer)
	return { class="ocean" }
end
```

and the following config, to write everything to the "ocean" layer, which does not have any `source_columns`

```
"ocean":{ "minzoom": 0,  "maxzoom": 14, "source": "coastline/water_polygons.shp", "simplify_below": 13, "simplify_level": 0.0001, "simplify_ratio": 2,  "index": true }
```

**Reply by systemed** [reply]
It does sound like a style issue. Maybe try debugging the tiles with vt2geojson first to check that the ocean objects are in there?

**Comment by BenOnTrack** [comment]
Sorry for my late feedback, I've been away the last 2 weeks.
I've found my issue - I wasn't used to shapefile and when de-zipping the source data https://osmdata.openstreetmap.de/download/water-polygons-split-4326.zip  I initially only kept the `.shp` while now I understand that I actually also need all the dependencies (.shx, .prj, .cpg, .dbf, ...) - it took me a while to figure that out! 

And I confirm now everything works as expected! 
Thanks again for your support, I'm learning every day!
Ben.

**Reply by systemed** [reply]
Excellent - glad it works!
